src/oc_erchef/apps/oc_chef_wm/src/chef_wm_actions_queue_monitoring.erl

```diff
@@ -0,0 +1,472 @@
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+
+%% This gen_server periodically checks the RabbitMQ management console for
+%% max_length and current queue length values.
+%%
+%% 0) If a max_length policy is not set on the exchange, do nothing and recheck
+%%    at the next interval.
+%% 1) If a max_length policy IS set on the exchange but a queue is not
+%%    bound to the /analytics exchange, do nothing and recheck at the
+%%    next interval.
+%% 2) If a max_length policy IS set on the exchange, store the value and
+%%    proceed to check the current queue length. Store max_length and
+%%    current length values.
+%% 2a) If current length < max_length, and the queue is < 80% full,
+%%     set queue_at_capacity = false, continue
+%% 2b) If current length < max_length, and the queue is >= 80% && < 100% full
+%%     log a warning message stating the current percentage of capacity,
+%%     set queue_at_capacity, continue
+%% 2b) If current length = max_length, set queue_at_capacity = true,
+%%     log a warning message stating the current percentage of capacity
+
+
+%% clients of chef_wm_actions_queue_monitoring use 2 API calls:
+%% is_queue_at_capacity/0 and message_dropped/0. When attempting to send
+%% message to RabbitMQ, a client can check is_queue_at_capacity/0 and
+%% optionally not send a message if true ("dropping a message"). Clients
+%% that drop messages should report that a message has been dropped
+%% to this gen_server via message_dropped/0. Note, message_dropped/0
+%% is a cast, while is_queue_at_capacity/0 is a call. If calls to
+%% is_queue_at_capacity/0 take "too long", a timeout can be set via
+%% the `rabbitmq_queue_length_timeout_millis` config setting. This
+%% setting will force is_queue_at_capacity/0 to return `true` upon
+%% timeout, even if it's not necessarily the queue that is timing
+%% out.
+
+
+-module(chef_wm_actions_queue_monitoring).
+
+-ifdef(TEST).
+-compile(export_all).
+-endif.
+
+-export([start_link/0]).
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3,
+         stop/0,
+         status/0,
+         is_queue_at_capacity/0,
+         message_dropped/0,
+         create_pool/0,
+         delete_pool/0,
+         get_pool_configs/0,
+         override_queue_at_capacity/1,
+         sync_check_current_state/0,
+         start_timer/0,
+         stop_timer/0
+
+        ]).
+
+-behaviour(gen_server).
+-define(SERVER, ?MODULE).
+
+
+-record(state, {
```

Comment:
It might be useful to give this state a descriptive prefix so that it can be recognized at a glance if it ever shows up in a stacktrace.